accepted/tail-calls.md

```diff
@@ -0,0 +1,1239 @@
+<!-- building consensus around the primary design questions and how they affect -->
+<!-- stakeholders. The fine details of a design will be finalized during -->
+<!-- implementation review. -->
+
+The proposal has two parts:
+
+1. Designing a new calling convention for Wasm that supports tail calls.
+2. Designing trampolines for Wasmtime that support calling into and out from
+   this new Wasm calling convention.
+
+Note that we *must* do (1) because the native calling conventions don't support
+general tail calls and that means that we *must* do (2) in turn since we will
+need trampolines to, at minimum, convert between different calling
+conventions. Given that we are doing both, however, the design of each is
+largely orthogonal.
+
+## New CLIF Instructions
+[proposal-instructions]: #new-clif-instructions
+
+First, we will introduce two new CLIF instructions: `return_call` and
+`return_call_indirect`. These are analogous to the Wasm tail calls proposal's
+instructions of the same name. These instructions are block terminators, the
+caller and callee function signatures must use the new Wasm calling convention
+(see below), and the caller and callee must return the same types. The
+`return_call` instruction has a static callee immediate while the
+`return_call_indirect` instruction takes a dynamic callee operand.
+
+## New Wasm Calling Conventions in Cranelift
+[proposal-calling-conventions]: #new-wasm-calling-conventions-in-cranelift
+
+We will define one new calling convention for each of our target
+architectures[^per-architecture]: `wasm_x64`, `wasm_aarch64`, and etc... The
+main difference from the native calling conventions will be that callees will
+clean up stack space for callers, regardless whether we are in a tail call or
+not. This is required because we need to clean up the stack arguments after
+exiting a function, and we no longer always exit a function by returning to the
+caller, we can also exit the function by tail calling to another function. In
+that new case, we can't delay cleaning up stack arguments until the last
+function in a tail call chain returns to the original caller. If every function
+in the tail call chain allocated stack arguments, this would lead to *O(n)*
+stack growth in the tail call chain, but we *must* maintain *O(1)* stack space
+for the whole chain. So callees must be responsible for cleaning up their own
+stack arguments.
+
+[^per-architecture]: Yes, one calling convention per-architecture not one
+    per-target. Currently Wasm compiled for x86-64 has a different calling
+    convention depending on if we are targeting Windows or Unix. That was
+    motivated by having the Wasm calling convention interoperable with the
+    host's native calling convention. Given that we have to support tail calls,
+    we can't have that interoperability anymore, so we can also switch to a
+    single calling convention on x86-64. Yay, fewer variants to mantain!
```

Comment:
They can have the same name on all architectures, right? Also isn't it still necessary to have a windows and non-windows version to account for any different redzone sizes (0 on x86_64 windows, 128 bytes I believe on x86_64 Unix) and unwinder impls? Those are part of the platform abi you aren't allowed to violate. Using a too big redzone will cause interrupts to mess with the stack and using the wrong unwinder impl breaks backtraces.